File: Clarity/src/main/java/org/cubexmc/clarity/ClarityConfig.java
package org.cubexmc.clarity;

import org.bukkit.configuration.file.FileConfiguration;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/** config.yml 的不可变快照。 */
public final class ClarityConfig {

    private static final List<String> DEFAULT_LEVELTOOLS_PDC_KEYS = List.of(
            "leveltools:leveltoolslevel",
            "leveltools:leveltoolsxp",
            "leveltools:leveltoolsreward"
    );

    private final boolean autoCleanOnJoin;
    private final boolean dryRun;
    private final long joinDelayTicks;
    private final List<String> attributeModifierIds;
    private final List<String> effectTypes;
    private final boolean effectsInfiniteOnly;
    private final boolean itemLevelToolsEnabled;
    private final List<String> itemLevelToolsPdcKeys;
    private final boolean itemLevelToolsRemoveLore;
    private final boolean itemLevelToolsRemoveAttributesOnMarkedItems;
    private final List<String> itemAttributeModifierIds;
    private final Map<String, Double> itemAttributeMaxAmounts;

    private ClarityConfig(boolean autoCleanOnJoin, boolean dryRun, long joinDelayTicks,
                          List<String> attributeModifierIds, List<String> effectTypes,
                          boolean effectsInfiniteOnly, boolean itemLevelToolsEnabled,
                          List<String> itemLevelToolsPdcKeys, boolean itemLevelToolsRemoveLore,
                          boolean itemLevelToolsRemoveAttributesOnMarkedItems,
                          List<String> itemAttributeModifierIds, Map<String, Double> itemAttributeMaxAmounts) {
        this.autoCleanOnJoin = autoCleanOnJoin;
        this.dryRun = dryRun;
        this.joinDelayTicks = joinDelayTicks;
        this.attributeModifierIds = List.copyOf(attributeModifierIds);
        this.effectTypes = List.copyOf(effectTypes);
        this.effectsInfiniteOnly = effectsInfiniteOnly;
        this.itemLevelToolsEnabled = itemLevelToolsEnabled;
        this.itemLevelToolsPdcKeys = List.copyOf(itemLevelToolsPdcKeys);
        this.itemLevelToolsRemoveLore = itemLevelToolsRemoveLore;
        this.itemLevelToolsRemoveAttributesOnMarkedItems = itemLevelToolsRemoveAttributesOnMarkedItems;
        this.itemAttributeModifierIds = List.copyOf(itemAttributeModifierIds);
        this.itemAttributeMaxAmounts = Map.copyOf(itemAttributeMaxAmounts);
    }

    public static ClarityConfig load(FileConfiguration c) {
        return new ClarityConfig(
                c.getBoolean("auto-clean-on-join", false),
                c.getBoolean("dry-run", true),
                Math.max(1L, c.getLong("join-delay-ticks", 40L)),
                c.getStringList("attributes.remove-modifier-ids"),
                c.getStringList("effects.remove-types"),
                c.getBoolean("effects.infinite-only", true),
                c.getBoolean("items.leveltools.enabled", true),
                stringListOrDefault(c, "items.leveltools.remove-pdc-keys", DEFAULT_LEVELTOOLS_PDC_KEYS),
                c.getBoolean("items.leveltools.remove-lore", true),
                c.getBoolean("items.leveltools.remove-attributes-on-marked-items", false),
                c.getStringList("items.attributes.remove-modifier-ids"),
                parseMaxAmounts(c.getStringList("items.attributes.max-amounts"))
        );
    }

    private static List<String> stringListOrDefault(FileConfiguration c, String path, List<String> fallback) {
        if (!c.contains(path)) {
            return fallback;
        }
        return c.getStringList(path);
    }

    private static Map<String, Double> parseMaxAmounts(List<String> entries) {
        Map<String, Double> out = new LinkedHashMap<>();
        for (String raw : entries) {
            if (raw == null) {
                continue;
            }
            String value = raw.trim();
            int split = value.lastIndexOf('=');
            if (split <= 0 || split >= value.length() - 1) {
                continue;
            }
            String key = value.substring(0, split).trim().toLowerCase(Locale.ROOT);
            if (key.isEmpty()) {
                continue;
            }
            try {
                double max = Double.parseDouble(value.substring(split + 1).trim());
                if (max >= 0.0D) {
                    out.put(key, max);
                }
            } catch (NumberFormatException ignored) {
                // Invalid config entries are ignored; scan output remains conservative.
            }
        }
        return out;
    }

    public boolean autoCleanOnJoin() {
        return autoCleanOnJoin;
    }

    public boolean dryRun() {
        return dryRun;
    }

    public long joinDelayTicks() {
        return joinDelayTicks;
    }

    public List<String> attributeModifierIds() {
        return attributeModifierIds;
    }

    public List<String> effectTypes() {
        return effectTypes;
    }

    public boolean effectsInfiniteOnly() {
        return effectsInfiniteOnly;
    }

    public boolean itemLevelToolsEnabled() {
        return itemLevelToolsEnabled;
    }

    public List<String> itemLevelToolsPdcKeys() {
        return itemLevelToolsPdcKeys;
    }

    public boolean itemLevelToolsRemoveLore() {
        return itemLevelToolsRemoveLore;
    }

    public boolean itemLevelToolsRemoveAttributesOnMarkedItems() {
        return itemLevelToolsRemoveAttributesOnMarkedItems;
    }

    public List<String> itemAttributeModifierIds() {
        return itemAttributeModifierIds;
    }

    public Map<String, Double> itemAttributeMaxAmounts() {
        return itemAttributeMaxAmounts;
    }
}

File: Clarity/src/main/kotlin/org/cubexmc/clarity/ClarityConfig.kt
package org.cubexmc.clarity

import java.util.LinkedHashMap
import java.util.Locale
import org.bukkit.configuration.file.FileConfiguration

/** config.yml 的不可变快照。 */
class ClarityConfig private constructor(
    private val autoCleanOnJoin: Boolean,
    private val dryRun: Boolean,
    private val joinDelayTicks: Long,
    attributeModifierIds: List<String>,
    effectTypes: List<String>,
    private val effectsInfiniteOnly: Boolean,
    private val itemLevelToolsEnabled: Boolean,
    itemLevelToolsPdcKeys: List<String>,
    private val itemLevelToolsRemoveLore: Boolean,
    private val itemLevelToolsRemoveAttributesOnMarkedItems: Boolean,
    itemAttributeModifierIds: List<String>,
    itemAttributeMaxAmounts: Map<String, Double>,
) {
    private val attributeModifierIds = java.util.List.copyOf(attributeModifierIds)
    private val effectTypes = java.util.List.copyOf(effectTypes)
    private val itemLevelToolsPdcKeys = java.util.List.copyOf(itemLevelToolsPdcKeys)
    private val itemAttributeModifierIds = java.util.List.copyOf(itemAttributeModifierIds)
    private val itemAttributeMaxAmounts = java.util.Map.copyOf(itemAttributeMaxAmounts)

    fun autoCleanOnJoin(): Boolean = autoCleanOnJoin
    fun dryRun(): Boolean = dryRun
    fun joinDelayTicks(): Long = joinDelayTicks
    fun attributeModifierIds(): List<String> = attributeModifierIds
    fun effectTypes(): List<String> = effectTypes
    fun effectsInfiniteOnly(): Boolean = effectsInfiniteOnly
    fun itemLevelToolsEnabled(): Boolean = itemLevelToolsEnabled
    fun itemLevelToolsPdcKeys(): List<String> = itemLevelToolsPdcKeys
    fun itemLevelToolsRemoveLore(): Boolean = itemLevelToolsRemoveLore
    fun itemLevelToolsRemoveAttributesOnMarkedItems(): Boolean = itemLevelToolsRemoveAttributesOnMarkedItems
    fun itemAttributeModifierIds(): List<String> = itemAttributeModifierIds
    fun itemAttributeMaxAmounts(): Map<String, Double> = itemAttributeMaxAmounts

    companion object {
        private val DEFAULT_LEVELTOOLS_PDC_KEYS = listOf(
            "leveltools:leveltoolslevel",
            "leveltools:leveltoolsxp",
            "leveltools:leveltoolsreward",
        )

        @JvmStatic
        fun load(config: FileConfiguration): ClarityConfig = ClarityConfig(
            config.getBoolean("auto-clean-on-join", false),
            config.getBoolean("dry-run", true),
            maxOf(1L, config.getLong("join-delay-ticks", 40L)),
            config.getStringList("attributes.remove-modifier-ids"),
            config.getStringList("effects.remove-types"),
            config.getBoolean("effects.infinite-only", true),
            config.getBoolean("items.leveltools.enabled", true),
            stringListOrDefault(config, "items.leveltools.remove-pdc-keys", DEFAULT_LEVELTOOLS_PDC_KEYS),
            config.getBoolean("items.leveltools.remove-lore", true),
            config.getBoolean("items.leveltools.remove-attributes-on-marked-items", false),
            config.getStringList("items.attributes.remove-modifier-ids"),
            parseMaxAmounts(config.getStringList("items.attributes.max-amounts")),
        )

        private fun stringListOrDefault(
            config: FileConfiguration,
            path: String,
            fallback: List<String>,
        ): List<String> = if (config.contains(path)) config.getStringList(path) else fallback

        private fun parseMaxAmounts(entries: List<String?>): Map<String, Double> {
            val result = LinkedHashMap<String, Double>()
            for (raw in entries) {
                val value = raw?.trim() ?: continue
                val split = value.lastIndexOf('=')
                if (split <= 0 || split >= value.length - 1) continue
                val key = value.substring(0, split).trim().lowercase(Locale.ROOT)
                if (key.isEmpty()) continue
                val max = value.substring(split + 1).trim().toDoubleOrNull() ?: continue
                if (max >= 0.0) result[key] = max
            }
            return result
        }
    }
}
